plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,476 @@
+
+import qualified Ledger
+import           Ledger.AddressMap                                 (AddressMap)
+import qualified Ledger.AddressMap                                 as AM
+import           Ledger.Constraints.OffChain                       (UnbalancedTx (..))
+import           Ledger.Slot                                       (Slot)
+import           Wallet.API                                        (signWithOwnPublicKey)
+import           Wallet.Effects                                    (ChainIndexEffect, SigningProcessEffect,
+                                                                    WalletEffect, ownPubKey, startWatching, submitTxn,
+                                                                    transactionConfirmed, walletSlot, watchedAddresses)
+
+import           Plutus.SCB.Command                                (saveBalancedTx, saveBalancedTxResult,
+                                                                    saveContractState, sendContractEvent)
+import           Plutus.SCB.Effects.Contract                       (ContractCommand (..), ContractEffect)
+import qualified Plutus.SCB.Effects.Contract                       as Contract
+import           Plutus.SCB.Effects.EventLog                       (EventLogEffect, runCommand, runGlobalQuery)
+import           Plutus.SCB.Effects.UUID                           (UUIDEffect, uuidNextRandom)
+import           Plutus.SCB.Events                                 (ChainEvent (..))
+import           Plutus.SCB.Events.Contract                        (ContractEvent (..), ContractInstanceId (..),
+                                                                    ContractInstanceState (..), ContractIteration (..),
+                                                                    ContractMailbox (..), ContractRequest (..),
+                                                                    ContractResponse (..), MailboxMessage (..),
+                                                                    PartiallyDecodedResponse (..), iterationZero)
+import qualified Plutus.SCB.Query                                  as Query
+import           Plutus.SCB.Types                                  (SCBError (..), Source (ContractEventSource, NodeEventSource, UserEventSource, WalletEventSource))
+import           Plutus.SCB.Utils                                  (render, tshow)
+
+import qualified Plutus.SCB.Core.Projections                       as Projections
+
+sendContractStateMessages ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => t
+    -> ContractInstanceId
+    -> ContractIteration
+    -> PartiallyDecodedResponse
+    -> Eff effs ()
+sendContractStateMessages t instanceID iteration newState = do
+    logInfo . render $ "Sending messages for contract" <+> pretty instanceID <+> "at iteration" <+> pretty iteration
+    let is = ContractInstanceState
```

Comment:
The idea of the iteration is to provide a way to tell which requests by a contract are current and which are stale.

You are right that this could be expressed as an aggregate, but with `sendContractStateMessages` we can have effects such as log/debugging messages, which we can't with the aggregate.

Reply:
Isn't this function enough for that? I am thinking that the `m` could be our Eff, so we can do whatever we want in there. I do realise that this may be a big change, though.

```haskell
getLatestStreamProjection :: (Monad m, Num position) => EventStoreReader key position m (StreamEvent key position event) -> StreamProjection key position state event -> m (StreamProjection key position state event)
```

Reply:
Hm that does look useful, thanks. I'll see if I can change it to that.

Reply:
The `m` is only used in the `EventStoreReader` (1st argument), not in the `Projection` (2nd argument) so we'd have to write an [`EventStoreReader`](https://hackage.haskell.org/package/eventful-core-0.2.0/docs/Eventful-Store-Class.html#t:EventStoreReader) that then uses the `ContractIteration` as key (if I read this right?) Anyway I think it's better to worry about performance later on, in a dedicated task. Otherwise this will take on the whole sprint.